File: psl_core/engine_v2/vision.py
"""Player vision system (Phase 2): IQ-linked field of view for pass perception."""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple, TYPE_CHECKING

from .physics import angle_between_points, angle_diff, distance, is_in_fov

if TYPE_CHECKING:
    from .player import Player
    from .config import EngineConfig


def compute_facing_direction(
    player_pos: Tuple[float, float],
    ball_pos: Tuple[float, float],
    last_action_target: Tuple[float, float] = None,
) -> float:
    """Compute player facing direction in degrees.

    Defaults to facing the ball. If the player just performed an action,
    they face toward the action target briefly.
    """
    target = last_action_target if last_action_target else ball_pos
    return angle_between_points(player_pos, target)


def compute_fov(iq: int, config: "EngineConfig") -> float:
    """Compute player field of view in degrees based on IQ.

    Average-IQ players use the base FOV; higher IQ widens the scan cone.
    """
    fov = config.vision_base_fov + (iq - 80) * config.vision_iq_bonus_factor
    return max(150.0, min(240.0, fov))


def compute_vision_distance(iq: int, config: "EngineConfig") -> float:
    """Compute how far a player can evaluate pass targets."""
    base = getattr(config, "vision_base_distance", 42.0)
    iq_bonus = getattr(config, "vision_iq_distance_bonus_factor", 0.35)
    max_distance = getattr(config, "vision_max_distance", 65.0)
    return max(24.0, min(max_distance, base + max(0, iq - 70) * iq_bonus))


def compute_player_facing(
    passer: "Player",
    attacking_right: bool,
) -> float:
    """Use live body orientation, with attack direction as fallback."""
    facing = getattr(passer, "facing_direction", 0.0)
    if abs(facing) > 1e-3:
        return facing
    return 0.0 if attacking_right else 180.0


@dataclass(frozen=True)
class VisionContext:
    """Per-pass perception envelope for candidate generation and trace."""

    facing: float
    fov: float
    half_fov: float
    max_distance: float

    def confidence(self, origin: Tuple[float, float], target: Tuple[float, float]) -> float:
        """Continuous perception confidence in [0, 1]."""
        d = distance(origin, target)
        if d <= 0.1:
            return 1.0
        target_angle = angle_between_points(origin, target)
        diff = abs(angle_diff(self.facing, target_angle))
        angle_conf = 1.0 - max(0.0, diff - self.half_fov * 0.65) / max(1.0, self.half_fov * 0.70)
        dist_conf = 1.0 - max(0.0, d - self.max_distance * 0.72) / max(1.0, self.max_distance * 0.45)
        return max(0.0, min(1.0, min(angle_conf, dist_conf)))

    def visible(self, origin: Tuple[float, float], target: Tuple[float, float]) -> bool:
        return self.confidence(origin, target) > 0.0


def build_vision_context(
    passer: "Player",
    config: "EngineConfig",
    attacking_right: bool,
) -> VisionContext:
    fov = compute_fov(passer.iq_value, config)
    return VisionContext(
        facing=compute_player_facing(passer, attacking_right),
        fov=fov,
        half_fov=fov / 2.0,
        max_distance=compute_vision_distance(passer.iq_value, config),
    )


def get_visible_targets(
    passer: "Player",
    teammates: List["Player"],
    ball_pos: Tuple[float, float],
    config: "EngineConfig",
) -> List["Player"]:
    """Filter teammates to only those within the passer's field of view.

    Args:
        passer: The player making the pass.
        teammates: All teammates.
        ball_pos: Current ball position for facing calculation.
        config: Engine configuration.

    Returns:
        List of teammates visible to the passer.
    """
    facing = compute_facing_direction(passer.pos, ball_pos)
    fov = compute_fov(passer.iq_value, config)
    half_fov = fov / 2.0

    visible = []
    for tm in teammates:
        if tm.index == passer.index:
            continue
        angle_to_tm = angle_between_points(passer.pos, tm.pos)
        if is_in_fov(facing, angle_to_tm, half_fov):
            visible.append(tm)

    # Safety: always see at least a couple of teammates (peripheral awareness)
    # If no one visible, allow all (to prevent stuck states)
    if not visible:
        return [tm for tm in teammates if tm.index != passer.index]

    return visible


def is_target_visible(
    passer: "Player",
    target_pos: Tuple[float, float],
    ball_pos: Tuple[float, float],
    config: "EngineConfig",
) -> bool:
    """Check if a specific target position is within the passer's FOV."""
    facing = compute_facing_direction(passer.pos, ball_pos)
    fov = compute_fov(passer.iq_value, config)
    half_fov = fov / 2.0
    angle_to_target = angle_between_points(passer.pos, target_pos)
    return is_in_fov(facing, angle_to_target, half_fov)

File: rust/engine_v2_core/src/vision.rs
use crate::physics::{angle_between_points, angle_diff, distance, is_in_fov};

#[derive(Clone, Copy, Debug)]
pub struct VisionContextInput {
    pub iq: f64,
    pub facing_direction: f64,
    pub attacking_right: bool,
    pub vision_base_fov: f64,
    pub vision_iq_bonus_factor: f64,
    pub vision_base_distance: f64,
    pub vision_iq_distance_bonus_factor: f64,
    pub vision_max_distance: f64,
}

#[derive(Clone, Copy, Debug)]
pub struct VisionContext {
    pub facing: f64,
    pub fov: f64,
    pub half_fov: f64,
    pub max_distance: f64,
}

pub fn compute_facing_direction(
    player_pos: (f64, f64),
    ball_pos: (f64, f64),
    last_action_target: Option<(f64, f64)>,
) -> f64 {
    angle_between_points(player_pos, last_action_target.unwrap_or(ball_pos))
}

pub fn compute_fov(iq: f64, vision_base_fov: f64, vision_iq_bonus_factor: f64) -> f64 {
    let fov = vision_base_fov + (iq - 80.0) * vision_iq_bonus_factor;
    fov.clamp(150.0, 240.0)
}

pub fn compute_vision_distance(
    iq: f64,
    vision_base_distance: f64,
    vision_iq_distance_bonus_factor: f64,
    vision_max_distance: f64,
) -> f64 {
    let distance = vision_base_distance + (iq - 70.0).max(0.0) * vision_iq_distance_bonus_factor;
    distance.clamp(24.0, vision_max_distance)
}

pub fn compute_player_facing(facing_direction: f64, _attacking_right: bool) -> f64 {
    facing_direction
}

impl VisionContext {
    pub fn confidence(&self, origin: (f64, f64), target: (f64, f64)) -> f64 {
        let d = distance(origin, target);
        if d <= 0.1 {
            return 1.0;
        }
        let target_angle = angle_between_points(origin, target);
        let diff = angle_diff(self.facing, target_angle).abs();
        let angle_conf =
            1.0 - (diff - self.half_fov * 0.65).max(0.0) / (self.half_fov * 0.70).max(1.0);
        let dist_conf =
            1.0 - (d - self.max_distance * 0.72).max(0.0) / (self.max_distance * 0.45).max(1.0);
        angle_conf.min(dist_conf).clamp(0.0, 1.0)
    }

    pub fn visible(&self, origin: (f64, f64), target: (f64, f64)) -> bool {
        self.confidence(origin, target) > 0.0
    }
}

pub fn build_vision_context(input: &VisionContextInput) -> VisionContext {
    let fov = compute_fov(
        input.iq,
        input.vision_base_fov,
        input.vision_iq_bonus_factor,
    );
    VisionContext {
        facing: compute_player_facing(input.facing_direction, input.attacking_right),
        fov,
        half_fov: fov / 2.0,
        max_distance: compute_vision_distance(
            input.iq,
            input.vision_base_distance,
            input.vision_iq_distance_bonus_factor,
            input.vision_max_distance,
        ),
    }
}

pub fn get_visible_target_indices(
    passer_index: usize,
    passer_pos: (f64, f64),
    passer_iq: f64,
    teammates: &[(usize, f64, f64)],
    ball_pos: (f64, f64),
    vision_base_fov: f64,
    vision_iq_bonus_factor: f64,
) -> Vec<usize> {
    let facing = compute_facing_direction(passer_pos, ball_pos, None);
    let half_fov = compute_fov(passer_iq, vision_base_fov, vision_iq_bonus_factor) / 2.0;
    let mut visible = Vec::new();
    let mut fallback = Vec::new();

    for (idx, x, y) in teammates {
        if *idx == passer_index {
            continue;
        }
        fallback.push(*idx);
        let angle_to_tm = angle_between_points(passer_pos, (*x, *y));
        if is_in_fov(facing, angle_to_tm, half_fov) {
            visible.push(*idx);
        }
    }

    if visible.is_empty() {
        fallback
    } else {
        visible
    }
}

pub fn is_target_visible(
    passer_pos: (f64, f64),
    passer_iq: f64,
    target_pos: (f64, f64),
    ball_pos: (f64, f64),
    vision_base_fov: f64,
    vision_iq_bonus_factor: f64,
) -> bool {
    let facing = compute_facing_direction(passer_pos, ball_pos, None);
    let half_fov = compute_fov(passer_iq, vision_base_fov, vision_iq_bonus_factor) / 2.0;
    let angle_to_target = angle_between_points(passer_pos, target_pos);
    is_in_fov(facing, angle_to_target, half_fov)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fov_and_distance_are_clamped_like_python() {
        assert_eq!(compute_fov(20.0, 180.0, 0.5), 150.0);
        assert_eq!(compute_fov(80.0, 180.0, 0.5), 180.0);
        assert_eq!(compute_fov(260.0, 180.0, 0.5), 240.0);
        assert_eq!(compute_vision_distance(30.0, 20.0, 0.35, 65.0), 24.0);
        assert_eq!(compute_vision_distance(120.0, 42.0, 0.35, 65.0), 59.5);
        assert_eq!(compute_vision_distance(200.0, 42.0, 0.35, 65.0), 65.0);
    }

    #[test]
    fn visible_targets_falls_back_when_none_are_in_fov() {
        let teammates = vec![(0, 0.0, 0.0), (1, -10.0, 5.0), (2, -12.0, -5.0)];
        let result =
            get_visible_target_indices(0, (0.0, 0.0), 80.0, &teammates, (10.0, 0.0), 180.0, 0.5);
        assert_eq!(result, vec![1, 2]);
    }

    #[test]
    fn zero_degree_is_a_valid_explicit_facing_direction() {
        assert_eq!(compute_player_facing(0.0, false), 0.0);
    }
}
